icefall/lm/rescore.py

```diff
@@ -0,0 +1,338 @@
+        sublist.
+
+    Returns:
+      Return a new ragged tensor, whose sublists either start with
+      or end with the given value.
+
+    >>> a = k2.RaggedTensor([[1, 3], [5]])
+    >>> a
+    [ [ 1 3 ] [ 5 ] ]
+    >>> concat(a, value=0, direction="left")
+    [ [ 0 1 3 ] [ 0 5 ] ]
+    >>> concat(a, value=0, direction="right")
+    [ [ 1 3 0 ] [ 5 0 ] ]
+
+    """
+    dtype = ragged.dtype
+    device = ragged.device
+
+    assert ragged.num_axes == 2, f"num_axes: {ragged.num_axes}"
+    pad_values = torch.full(
+        size=(ragged.tot_size(0), 1),
+        fill_value=value,
+        device=device,
+        dtype=dtype,
+    )
+    pad = k2.RaggedTensor(pad_values)
+
+    if direction == "left":
+        ans = k2.ragged.cat([pad, ragged], axis=1)
+    elif direction == "right":
+        ans = k2.ragged.cat([ragged, pad], axis=1)
+    else:
+        raise ValueError(
+            f'Unsupported direction: {direction}. " \
+            "Expect either "left" or "right"'
+        )
+    return ans
+
+
+def add_bos(ragged: k2.RaggedTensor, bos_id: int) -> k2.RaggedTensor:
+    """Add BOS to each sublist.
+
+    Args:
+      ragged:
+        A ragged tensor with two axes.
+      bos_id:
+        The ID of the BOS symbol.
+
+    Returns:
+      Return a new ragged tensor, where each sublist starts with BOS.
+
+    >>> a = k2.RaggedTensor([[1, 3], [5]])
+    >>> a
+    [ [ 1 3 ] [ 5 ] ]
+    >>> add_bos(a, bos_id=0)
+    [ [ 0 1 3 ] [ 0 5 ] ]
+
+    """
+    return concat(ragged, bos_id, direction="left")
+
+
+def add_eos(ragged: k2.RaggedTensor, eos_id: int) -> k2.RaggedTensor:
+    """Add EOS to each sublist.
+
+    Args:
+      ragged:
+        A ragged tensor with two axes.
+      bos_id:
+        The ID of the EOS symbol.
+
+    Returns:
+      Return a new ragged tensor, where each sublist ends with EOS.
+
+    >>> a = k2.RaggedTensor([[1, 3], [5]])
+    >>> a
+    [ [ 1 3 ] [ 5 ] ]
+    >>> add_eos(a, eos_id=0)
+    [ [ 1 3 0 ] [ 5 0 ] ]
+
+    """
+    return concat(ragged, eos_id, direction="right")
+
+
+def make_hyp_to_ref_map(row_splits: torch.Tensor) -> torch.Tensor:
+    """
+    TODO: Add documentation
+
+    >>> row_splits = torch.tensor([0, 3, 5], dtype=torch.int32)
+    >>> make_hyp_to_ref_map(row_splits)
+    tensor([0, 0, 0, 3, 3], dtype=torch.int32)
+    """
+    device = row_splits.device
+    sizes = (row_splits[1:] - row_splits[:-1]).tolist()
+    offsets = row_splits[:-1]
+
+    map_tensor_list = []
+    for size, offset in zip(sizes, offsets):
+        # Explanation of the following operations
+        # assume size is 3, offset is 2
+        # torch.zeros() + offset is [2, 2, 2]
+        map_tensor = (
+            torch.zeros(size, dtype=torch.int32, device=device) + offset
+        )
+        map_tensor_list.append(map_tensor)
+
+    return torch.cat(map_tensor_list)
+
+
+def compute_alignment(
+    tokens: k2.RaggedTensor,
+    shape: k2.RaggedShape,
+) -> k2.Fsa:
```

Comment:
I think it would be especially helpful to have some documentation for this, esp. about the input and output formats.  Would be easier to follow what the function is doing.  Same with make_hyp_to_ref_map.